include/sys/arc.h

```diff
@@ -63,8 +63,15 @@ extern "C" {
 	(hdr)->b_psize = ((x) >> SPA_MINBLOCKSHIFT); \
 } while (0)
 
+/* The asize in the header is only used by L2 cache */
+#define	HDR_SET_ASIZE(hdr, x) do { \
+	ASSERT(IS_P2ALIGNED((x), 1U << SPA_MINBLOCKSHIFT)); \
+	(hdr)->b_asize = ((x) >> SPA_MINBLOCKSHIFT); \
+} while (0)
+
 #define	HDR_GET_LSIZE(hdr)	((hdr)->b_lsize << SPA_MINBLOCKSHIFT)
 #define	HDR_GET_PSIZE(hdr)	((hdr)->b_psize << SPA_MINBLOCKSHIFT)
+#define	HDR_GET_ASIZE(hdr)	((hdr)->b_asize << SPA_MINBLOCKSHIFT)
```

Comment:
I find confusing naming it `ASIZE` (both here and in the structure), since it is specific only for L2ARC.  We can not have non-L2ARC asize here since due to multiple block copies on different vdevs they might have different asize's.  It is just because we do not allow multiple L2ARC copies it makes sense here. I'm thinking about something more specific, like `L2SIZE` and `b_l2size`.  It would be nice for it to reside in `l2arc_buf_hdr_t`, but it does not worth the memory waste.

Reply:
good point, I'll update the names